### PR TITLE
Add material with specializations for debugging meshes

### LINKED ISSUE #4235: Add PBR debug render modes

### PR DESCRIPTION
**Pull request by torsteingrindvik**
# Objective

- Fixes #4235 
- Supercedes #5398 

Allows users to debug meshes by looking at UVs, and world {position, normals, tangents}.

## Solution

Add a material which allows per-mesh viewing of the above attributes.

### Added

- Added `DebugMeshMaterial` with pipeline specializations for attributes: UVs, world position, world normals, world tangents.
- Added `debug_mesh` example to show usage of `DebugMeshMaterial`s on a simple cube and on the flight helmet.

#### UVs (2nd debug helmet front-to-back)

https://user-images.githubusercontent.com/52322338/215233129-99703c21-d7fa-4dc4-bea2-5ad06c2d8c5e.mp4

Compare the above Bevy video to the glTF Sample Viewer (see [here](https://github.khronos.org/glTF-Sample-Viewer-Release/)):

bilde

This generally looks correct to me although the web viewer has darker colors.

### Features

Perhaps local position, normals, tangents would be good to have.
Let me know. I haven't looked into if I have all the functions/data I need in the frag shader for that to be a simple calculation or if a different vertex shader would be needed then.

### CODE REVIEW

File: crates/bevy_pbr/src/render/debug_mesh.wgsl [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#endif
+
+#ifdef DEBUG_UVS
+#ifdef VERTEX_UVS
+    // Modulo to show how the UVs would tile a texture
+    // if the related sampler was set to repeat.
+    // This is a more useful default than to saturate/clamp.
+    return vec4<f32>(uv.x % 1.0, uv.y % 1.0, 0.0, 1.0);
+#else
+    return vec4<f32>(0.0, 0.0, 0.0, 1.0);
```
Comment (torsteingrindvik) [inline crates/bevy_pbr/src/render/debug_mesh.wgsl:20]: What is a sane default when a user wants to debug UVs (or other attributes) but there are none?
I opted for just showing black.

File: Cargo.toml [hunk 2]
```diff
@@ -13,7 +13,11 @@ repository = "https://github.com/bevyengine/bevy"
 rust-version = "1.67.0"
 
 [workspace]
-exclude = ["benches", "crates/bevy_ecs_compile_fail_tests", "crates/bevy_reflect_compile_fail_tests"]
+exclude = [
```
Comment (torsteingrindvik) [inline Cargo.toml:16]: Just a format-on-save whitespace change.

File: crates/bevy_pbr/src/render/debug_mesh.wgsl (hunk 1 above)
Comment (alice-i-cecile) [reply crates/bevy_pbr/src/render/debug_mesh.wgsl:20]: I would expect something garish and visibly different: I would go with hot pink as that's pretty common for "something has gone wrong".

File: Cargo.toml [hunk 3]
```diff
@@ -347,6 +351,16 @@ description = "Illustrates bloom configuration using HDR and emissive materials"
...
 
+[[example]]
+name = "debug_mesh"
+path = "examples/3d/debug_mesh.rs"
+
+[package.metadata.example.debug_mesh]
+name = "Debug mesh"
+description = "Debug meshes by displaying UVs, world normals, etc."
+category = "3D Rendering"
+wasm = false
```
Comment (torsteingrindvik) [inline Cargo.toml:362]: `wasm = false` is this an opt-out of having the example on Bevy's homepage?
I haven't tried the example on a wasm build, let me know what to put here.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Looks awesome! Can we enable this as a mode in our scene_viewer tool?

File: Cargo.toml (hunk 3 above)
Comment (alice-i-cecile) [reply Cargo.toml:362]: That's correct. I'd set this to `true`, it's visually impressive and good to verify.

**Comment by superdump** [comment]
I did a PR a long time ago that focused more on screen-wide PBR debug stuff like this. https://github.com/bevyengine/bevy/pull/5398 I'm not sure which I feel is better, or if both are desirable. I feel like a whole-screen solution is useful, just as a per-mesh entity solution is useful. Perhaps they can share some parts or something. Also things have moved on a lot since my PR. It needs updating and there are many more things to debug now. :) Would you perhaps be up for putting together a combined solution?

**Comment by SneakyBerry** [comment]
Is it technically possible to represent UV map like this?
image

**Comment by torsteingrindvik** [comment]
@alice-i-cecile thanks for the review comments, I'll address them after figuring out where to take this.

@superdump Locally now I've merged your PR into this one (so git history should be preserved). Adapted it a bit since there have been a few changes, was not too much work. So now the scene viewer works when cycling all of the enum variants from your PR.

Since yours was a lot more thorough it makes sense to use what you made.

I'm thinking I will try to do something similar to the `WireframePlugin` where it's possible to both have a global setting and a per-mesh setting.

Another thing: If we have the scene viewer, do we still want the example I added or should I remove it?

File: crates/bevy_pbr/src/render/debug_mesh.wgsl (hunk 1 above)
Comment (torsteingrindvik) [reply crates/bevy_pbr/src/render/debug_mesh.wgsl]: Using @superdump's shader now, which does indeed use pink.

File: Cargo.toml (hunk 3 above)
Comment (torsteingrindvik) [reply Cargo.toml]: Set to true now.

**Comment by torsteingrindvik** [comment]
@alice-i-cecile Merged in @superdump's PR so the scene viewer is included now.

@superdump From your PR the main changes I've made is to update the PBR functions shader slightly since the `prepare_normals` function was replace, and also I made `PbrDebug` an optional resource instead of having a `PbrDebug::None` variant.

To attempt getting the best of both worlds I added `PbrDebugMaterial` to allow per-mesh choices in the same way I did in this PR originally, but it's all pink now even though the shader defs are added:
bilde

So I'll debug that.
However the scene viewer works:
bilde


Also even though I did a `git merge superdump/<branch>` directly it did not preserve you as author @superdump  😢 
Not sure why. Let me know if there's some git magic I can do here.

**Comment by superdump** [comment]
> ~I'm thinking I will try to do something similar to the `WireframePlugin` where it's possible to both have a global setting and a per-mesh setting.~ (EDIT: I'll push to this PR with all of this when I figure that out. If so this PR will combine both of ours. Hopefully that's ok) (EDIT 2: I realize the WireframePlugin does not _replace_ meshes, it just draws on top. So that method is not applicable I think)

That's fine, just retain attribution (i.e. add my stuff in its own commit and make me the author of the commit with `git commit --author="Robert Swain <robert.swain@gmail.com>"`)

> Another thing: If we have the scene viewer, do we still want the example I added or should I remove it?

I think having an example is good too. The `scene_viewer`, being a tool, is perhaps not the first place to think to look, whereas the examples are a de facto go to to look up how to do stuff.

**Comment by superdump** [comment]
This is super cool! The ability to have multiple instances of a mesh side-by-side with different debug views at the same time is surely incredibly useful for hacking away on shaders!

**Comment by superdump** [comment]
Could you add:
- Fixes #4235 
- Supercedes #5398 

into the PR description?

**Comment by superdump** [comment]
Merging #7431 so you can build on top of that now to simplify the ifdeffery in this one. :)

**Comment by torsteingrindvik** [comment]
Thanks! I'll revisit this in the next couple of days and apply that PR 👍

**Comment by torsteingrindvik** [comment]
Got rid of a lot of `#endif`s in the `pbr_debug.wgsl` shader now, so that's in place.

However I'm a bit stuck on understanding why the material approach does not work (it works only for the few shown in the images).

I'll compare my understanding of the global/resources vs. the material approach:

The global/resource based version checks if the `PbrDebug` resource changes, if so set this on the mesh pipeline.
Then invalidate the `SpecializedMeshPipelines` cache for `MaterialPipeline<StandardMaterial>`. I'm not sure I understand this entirely, but I think it's just because the _key_ must change before a new pipeline is created, and the `PbrDebug` field is not part of the key.
(Please let me know if it's more nuanced).
And then with the new pipeline in action, the proper shader definitions are added from `PbrDebug`.

By this logic I created `PbrDebugMaterial`, and used `PbrDebug` as the `bind_group_data`, which lets me specialize this material on which variant of `PbrDebug` is used.
The shader defs are added in the specialization of the material.
Since now the `PbrDebug` is part of the key, each variant used should result in a new custom pipeline, and therefore no cache invalidation is necessary here.

So intuitively I thought both approaches should work since the goal is simply to get the right shader defs in place.
Any insight on why this is not the case would be very helpful.

**Comment by superdump** [comment]
At least at first glance, your description sounds reasonable. I'll take a look at the PR locally.

**Comment by superdump** [comment]
The problem is that the `PbrDebugMaterial` is trying to use the main PBR shader without using any of the `StandardMaterial` data nor shader defs. That was simpler for the way I did it because that I didn't have to deal with materials as I managed the shader defs globally. Hmm, we really need a good way of extending / modifying the `StandardMaterial` without copying it, modifying it, and then maintaining that.

To be more specific and concrete, `TangentSpaceNormalMap` for example needs to have vertex tangents, vertex normals, vertex uvs, and a normal map texture. I noticed that `STANDARDMATERIAL_NORMAL_MAP` was missing and then realised that the `PbrDebugMaterial` only contains `PbrDebug` and none of the `StandardMaterial` bindings.

Hmm, I think to avoid duplication, the material approach will need some more thought. I'm not yet sure how to solve it without just building it into `StandardMaterial`, which we could do...

**Comment by torsteingrindvik** [comment]
Thanks for looking into that.

Since we currently don't have a clear image of how to solve that, should I perhaps split this PR up?
I could rebase and keep your PR here (`PbrDebug` and scene viewer stuff), then split the `PbrDebugMaterial` related stuff into a draft PR until we know more.
Such that we have something to merge before 0.10.

**Comment by alice-i-cecile** [comment]
I don't think this is close enough to shippable to make the 0.10 release, but feel free to prove me wrong ;)

**Comment by torsteingrindvik** [comment]
I see the PBR shaders have changed a lot, seems the PBR pipeline has gotten more structured and extensible in a year which is nice.
But I don't have a good overview of the pipeline now so I'll close- I think a re-do of this PR would likely be a lot better.
(If anyone sees this and would like to continue from a rebase instead of from scratch of course feel free to.)

**Pull request closed without merging**